File: src/main/java/io/github/tamawish/pureeconomy/command/PayCommand.java
package io.github.tamawish.pureeconomy.command;

import io.github.tamawish.pureeconomy.PureEconomy;
import io.github.tamawish.pureeconomy.economy.Currency;
import io.github.tamawish.pureeconomy.economy.EconomyService;
import io.github.tamawish.pureeconomy.lang.Lang;
import io.github.tamawish.pureeconomy.util.Amounts;
import org.bukkit.Bukkit;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabCompleter;
import org.bukkit.entity.Player;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.UUID;

public final class PayCommand implements CommandExecutor, TabCompleter {

    private final PureEconomy plugin;

    public PayCommand(PureEconomy plugin) {
        this.plugin = plugin;
        plugin.getCommand("pay").setTabCompleter(this);
    }

    @Override
    public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
        Lang lang = plugin.lang();
        if (!(sender instanceof Player player)) {
            lang.send(sender, "player-only");
            return true;
        }
        if (!player.hasPermission("pureeconomy.pay")) {
            lang.send(sender, "no-permission");
            return true;
        }
        if (args.length < 2) {
            lang.send(sender, "usage-pay");
            return true;
        }

        EconomyService eco = plugin.economy();
        UUID target = eco.resolve(args[0]);
        if (target == null) {
            lang.send(sender, "unknown-player", Lang.of("player", args[0]));
            return true;
        }
        if (target.equals(player.getUniqueId())) {
            lang.send(sender, "pay-self");
            return true;
        }

        BigDecimal amount = Amounts.parse(args[1]);
        if (amount == null) {
            lang.send(sender, "invalid-amount");
            return true;
        }

        Currency currency = args.length >= 3 ? eco.currency(args[2]) : eco.defaultCurrency();
        if (currency == null) {
            lang.send(sender, "unknown-currency", Lang.of("currency", args.length >= 3 ? args[2] : ""));
            return true;
        }
        if (!currency.payable()) {
            lang.send(sender, "pay-disabled");
            return true;
        }

        BigDecimal min = Amounts.parse(plugin.getConfig().getString("pay-minimum", "0.01"));
        if (min != null && amount.compareTo(min) < 0) {
            lang.send(sender, "pay-minimum", Lang.of("amount", currency.format(min)));
            return true;
        }

        amount = currency.normalize(amount);
        if (!eco.has(player.getUniqueId(), currency, amount)) {
            lang.send(sender, "not-enough", Lang.of("currency", currency.name()));
            return true;
        }

        BigDecimal next = eco.get(target, currency).add(amount);
        if (currency.exceedsMax(next)) {
            lang.send(sender, "max-balance", Lang.of("currency", currency.name()));
            return true;
        }

        if (!eco.transfer(player.getUniqueId(), target, currency, amount)) {
            lang.send(sender, "not-enough", Lang.of("currency", currency.name()));
            return true;
        }

        String pretty = currency.format(amount);
        lang.send(player, "pay-sent", Lang.of("amount", pretty, "player", eco.nameOf(target)));
        Player online = Bukkit.getPlayer(target);
        if (online != null) {
            lang.send(online, "pay-received", Lang.of("amount", pretty, "player", player.getName()));
        }
        return true;
    }

    @Override
    public List<String> onTabComplete(CommandSender sender, Command command, String alias, String[] args) {
        if (args.length == 1) {
            List<String> names = new ArrayList<>();
            plugin.getServer().getOnlinePlayers().forEach(p -> names.add(p.getName()));
            return filter(names, args[0]);
        }
        if (args.length == 3) {
            return filter(plugin.economy().currencyIds(), args[2]);
        }
        return Collections.emptyList();
    }

    private static List<String> filter(List<String> in, String prefix) {
        String p = prefix.toLowerCase();
        List<String> out = new ArrayList<>();
        for (String s : in) {
            if (s.toLowerCase().startsWith(p)) {
                out.add(s);
            }
        }
        return out;
    }
}

File: src/main/kotlin/io/github/tamawish/pureeconomy/command/PayCommand.kt
package io.github.tamawish.pureeconomy.command

import io.github.tamawish.pureeconomy.PureEconomy
import io.github.tamawish.pureeconomy.api.EconomyStatus
import io.github.tamawish.pureeconomy.lang.Lang
import io.github.tamawish.pureeconomy.permission.Permissions.Node
import io.github.tamawish.pureeconomy.util.Amounts
import io.github.tamawish.pureeconomy.util.CommandCompletions
import io.github.tamawish.pureeconomy.util.Schedulers
import org.bukkit.Bukkit
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.command.TabCompleter
import org.bukkit.entity.Player

/** Handles `/pay` transfers between player wallets. */
class PayCommand(
    private val plugin: PureEconomy,
) : CommandExecutor,
    TabCompleter {
    override fun onCommand(
        sender: CommandSender,
        command: Command,
        label: String,
        args: Array<out String>,
    ): Boolean {
        val lang = plugin.lang()
        if (sender !is Player) {
            lang.send(sender, "player-only")
            return true
        }
        if (!plugin.permissions().has(sender, Node.PAY)) {
            lang.send(sender, "no-permission")
            return true
        }
        if (args.size < 2) {
            lang.send(sender, "usage-pay")
            return true
        }

        val eco = plugin.economy()
        val target = eco.resolve(args[0], lookupOffline = false)
        if (target == null) {
            lang.send(sender, "unknown-player", Lang.of("player", args[0]))
            return true
        }
        if (target == sender.uniqueId) {
            lang.send(sender, "pay-self")
            return true
        }

        var amount = Amounts.parse(args[1])
        if (amount == null) {
            lang.send(sender, "invalid-amount")
            return true
        }

        val currency = if (args.size >= 3) eco.currency(args[2]) else eco.defaultCurrency()
        if (currency == null) {
            lang.send(sender, "unknown-currency", Lang.of("currency", if (args.size >= 3) args[2] else ""))
            return true
        }
        if (!currency.payable()) {
            lang.send(sender, "pay-disabled")
            return true
        }

        val min = Amounts.parse(plugin.settings().getString("pay-minimum", "0.01"))
        if (min != null && amount < min) {
            lang.send(sender, "pay-minimum", Lang.of("amount", currency.format(min)))
            return true
        }

        amount = currency.normalize(amount)
        val paid = eco.transferResult(sender.uniqueId, target, currency, amount)
        when (paid.status) {
            EconomyStatus.INSUFFICIENT -> {
                lang.send(sender, "not-enough", Lang.of("currency", currency.name()))
                return true
            }
            EconomyStatus.MAX_BALANCE -> {
                lang.send(sender, "max-balance", Lang.of("currency", currency.name()))
                return true
            }
            EconomyStatus.SUCCESS -> {}
            else -> {
                lang.send(sender, "not-enough", Lang.of("currency", currency.name()))
                return true
            }
        }

        val pretty = currency.format(amount)
        lang.send(sender, "pay-sent", Lang.of("amount", pretty, "player", eco.nameOf(target)))
        val online = Bukkit.getPlayer(target)
        if (online != null) {
            val payerName = sender.name
            // Recipient may be on another Folia region — schedule on their entity thread.
            Schedulers.runAtEntity(plugin, online) {
                lang.send(online, "pay-received", Lang.of("amount", pretty, "player", payerName))
            }
        }
        return true
    }

    override fun onTabComplete(
        sender: CommandSender,
        command: Command,
        alias: String,
        args: Array<out String>,
    ): List<String> {
        if (args.size == 1) {
            val names = ArrayList<String>()
            plugin.server.onlinePlayers.forEach { names.add(it.name) }
            return CommandCompletions.filter(names, args[0])
        }
        if (args.size == 3) {
            return CommandCompletions.filter(plugin.economy().currencyIds(), args[2])
        }
        return emptyList()
    }
}
